django_setup_configuration/base.py

```diff
@@ -0,0 +1,168 @@
+from django.db.models.fields.related import OneToOneField
+
+from .constants import basic_field_description
+
+
+@dataclass(frozen=True, slots=True)
+class ConfigField:
+    name: str
+    verbose_name: str
+    description: str
+    default_value: str
+    field_description: str
+
+
+@dataclass
+class Fields:
+    all: set[ConfigField] = field(default_factory=set)
+    required: set[ConfigField] = field(default_factory=set)
+
+
+class ConfigSettingsModel:
+    model: models.Model
+    display_name: str
+    namespace: str
+    required_fields = tuple()
+    all_fields = tuple()
+    excluded_fields = ("id",)
+
+    def __init__(self):
+        self.config_fields = Fields()
+
+        self.create_config_fields(
+            require=self.required_fields,
+            exclude=self.excluded_fields,
+            include=self.all_fields,
+            model=self.model,
+        )
+
+    @classmethod
+    def get_setting_name(cls, field: ConfigField) -> str:
+        return f"{cls.namespace}_" + field.name.upper()
+
+    @staticmethod
+    def get_default_value(field: models.Field) -> str:
+        default = field.default
+
+        if default is NOT_PROVIDED:
+            return "No default"
+
+        # needed to make `generate_config_docs` idempotent
+        # because UUID's are randomly generated
+        if isinstance(field, models.UUIDField):
+            return "random UUID string"
+
+        # if default is a function, call the function to retrieve the value;
+        # we don't immediately return because we need to check the type first
+        # and cast to another type if necessary (e.g. list is unhashable)
+        if callable(default):
+            default = default()
+
+        if isinstance(default, Mapping):
+            return str(default)
+
+        # check for field type as well to avoid splitting values from CharField
+        if isinstance(field, (JSONField, ArrayField)) and isinstance(default, Sequence):
+            try:
+                return ", ".join(str(item) for item in default)
+            except TypeError:
+                return str(default)
+
+        return default
+
+    @staticmethod
+    def get_field_description(field: models.Field) -> str:
+        # fields with choices
+        if choices := field.choices:
+            example_values = [choice[0] for choice in choices]
+            return ", ".join(example_values)
+
+        field_type = type(field)
+        match field_type:
+            case item if item in basic_field_description.keys():
+                return basic_field_description.get(item)
+            case _:
+                return "No information available"
+
+    def get_concrete_model_fields(self, model) -> Iterator[models.Field]:
+        """
+        Get all concrete fields for a given `model`, skipping over backreferences like
+        `OneToOneRel` and fields that are blacklisted
+        """
+        return (
+            field
+            for field in model._meta.concrete_fields
```

Comment:
`get_fields()` returns not only `CharField`, `IntergerField` and the like, but also things like `OneToOneRel`, which we need to skip to avoid recursion errors.

Reply:
But you can just filter on fields used in the config step, right?